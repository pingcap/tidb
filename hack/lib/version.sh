#!/bin/bash
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# -----------------------------------------------------------------------------
# Version management helpers.  These functions help to set, save and load the
# following variables:
#
#    TIDB_GIT_COMMIT - The git commit id corresponding to this
#          source code.
#    TIDB_GIT_TREE_STATE - "clean" indicates no changes since the git commit id
#        "dirty" indicates source code changes after the git commit id
#    TIDB_GIT_VERSION - "vX.Y" used to indicate the last release version.

# Grovels through git to set a set of env variables.
#
# If TIDB_GIT_VERSION_FILE, this function will load from that file instead of
# querying git.
tidb::version::get_version_vars() {
  local git=(git --work-tree "${TIDB_ROOT}")

  if [[ -n ${TIDB_GIT_COMMIT-} ]] || TIDB_GIT_COMMIT=$("${git[@]}" rev-parse "HEAD^{commit}" 2>/dev/null); then
    if [[ -z ${TIDB_GIT_TREE_STATE-} ]]; then
      # Check if the tree is dirty.  default to dirty
      if git_status=$("${git[@]}" status --porcelain 2>/dev/null) && [[ -z ${git_status} ]]; then
        TIDB_GIT_TREE_STATE="clean"
      else
        TIDB_GIT_TREE_STATE="dirty"
      fi
    fi

    # Use git describe to find the version based on annotated tags.
    if [[ -n ${TIDB_GIT_VERSION-} ]] || TIDB_GIT_VERSION=$("${git[@]}" describe --tags --abbrev=14 "${TIDB_GIT_COMMIT}^{commit}" 2>/dev/null); then
      # This translates the "git describe" to an actual semver.org
      # compatible semantic version that looks something like this:
      #   v1.1.0-alpha.0.6+84c76d1142ea4d
      #
      # TODO: We continue calling this "git version" because so many
      # downstream consumers are expecting it there.
      DASHES_IN_VERSION=$(echo "${TIDB_GIT_VERSION}" | sed "s/[^-]//g")
      if [[ "${DASHES_IN_VERSION}" == "---" ]] ; then
        # We have distance to subversion (v1.1.0-subversion-1-gCommitHash)
        TIDB_GIT_VERSION=$(echo "${TIDB_GIT_VERSION}" | sed "s/-\([0-9]\{1,\}\)-g\([0-9a-f]\{14\}\)$/.\1\+\2/")
      elif [[ "${DASHES_IN_VERSION}" == "--" ]] ; then
        # We have distance to base tag (v1.1.0-1-gCommitHash)
        TIDB_GIT_VERSION=$(echo "${TIDB_GIT_VERSION}" | sed "s/-g\([0-9a-f]\{14\}\)$/+\1/")
      fi
      if [[ "${TIDB_GIT_TREE_STATE}" == "dirty" ]]; then
        # git describe --dirty only considers changes to existing files, but
        # that is problematic since new untracked .go files affect the build,
        # so use our idea of "dirty" from git status instead.
        TIDB_GIT_VERSION+="-dirty"
      fi
    fi
  fi
}

# Saves the environment flags to $1
tidb::version::save_version_vars() {
  local version_file=${1-}
  [[ -n ${version_file} ]] || {

      echo "!!! Internal error.  No file specified in tidb::version::save_version_vars"
    return 1
  }

  cat <<EOF >"${version_file}"
TIDB_GIT_COMMIT='${TIDB_GIT_COMMIT-}'
TIDB_GIT_TREE_STATE='${TIDB_GIT_TREE_STATE-}'
TIDB_GIT_VERSION='${TIDB_GIT_VERSION-}'
TIDB_GIT_MAJOR='${TIDB_GIT_MAJOR-}'
TIDB_GIT_MINOR='${TIDB_GIT_MINOR-}'
EOF
}

# Loads up the version variables from file $1
tidb::version::load_version_vars() {
  local version_file=${1-}
  [[ -n ${version_file} ]] || {
    echo "!!! Internal error.  No file specified in tidb::version::load_version_vars"
    return 1
  }

  source "${version_file}"
}

tidb::version::ldflag() {
  local key=${1}
  local val=${2}

  # If you update these, also update the list pkg/version/def.bzl.
  echo "-X ${TIDB_GO_PACKAGE}/version.${key}=${val}"
}

# Prints the value that needs to be passed to the -ldflags parameter of go build
# in order to set the TiDB based on the git tree status.
# IMPORTANT: if you update any of these, also update the lists in
# pkg/version/def.bzl and hack/print-workspace-status.sh.
tidb::version::ldflags() {
  tidb::version::get_version_vars

  local -a ldflags=($(tidb::version::ldflag "buildDate" "$(date -u +'%Y-%m-%dT%H:%M:%SZ')"))
  if [[ -n ${TIDB_GIT_COMMIT-} ]]; then
    ldflags+=($(tidb::version::ldflag "gitCommit" "${TIDB_GIT_COMMIT}"))
    ldflags+=($(tidb::version::ldflag "gitTreeState" "${TIDB_GIT_TREE_STATE}"))
  fi

  if [[ -n ${TIDB_GIT_VERSION-} ]]; then
    ldflags+=($(tidb::version::ldflag "gitVersion" "${TIDB_GIT_VERSION}"))
  fi

  # The -ldflags parameter takes a single string, so join the output.
  echo "${ldflags[*]-}"
}
