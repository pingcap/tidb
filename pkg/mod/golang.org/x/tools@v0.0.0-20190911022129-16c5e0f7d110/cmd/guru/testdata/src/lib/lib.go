package lib

type Type int

func (Type) Method(x *int) *int {
	return x
}

func Func() {
}

const Const = 3

var Var = 0

type Sorter interface {
	Len() int
	Less(i, j int) bool
	Swap(i, j int)
}

type Outer struct {
	A int
	b int
	inner
}

type inner struct {
	C bool
	d string
	recursive
}

type recursive struct {
	E bool
	*inner
}
