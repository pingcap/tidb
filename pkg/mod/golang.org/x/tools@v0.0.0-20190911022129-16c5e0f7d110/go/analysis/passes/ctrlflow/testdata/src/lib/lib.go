package lib

func CanReturn() {}

func NoReturn() {
	for {
	}
}
