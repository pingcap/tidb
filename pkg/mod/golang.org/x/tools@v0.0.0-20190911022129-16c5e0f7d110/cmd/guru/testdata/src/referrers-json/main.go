package main

// Tests of 'referrers' query.
// See go.tools/guru/guru_test.go for explanation.
// See referrers.golden for expected query results.

import "lib"

type s struct {
	f int
}

func main() {
	var v lib.Type = lib.Const // @referrers ref-package "lib"
	_ = v.Method               // @referrers ref-method "Method"
	_ = v.Method
	v++ //@referrers ref-local "v"
	v++

	_ = s{}.f // @referrers ref-field "f"

	var s2 s
	s2.f = 1
}
