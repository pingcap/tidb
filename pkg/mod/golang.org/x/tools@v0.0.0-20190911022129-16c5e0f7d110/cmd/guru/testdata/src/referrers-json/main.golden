-------- @referrers ref-package --------
{
	"desc": "package lib"
}
{
	"package": "definition-json",
	"refs": [
		{
			"pos": "testdata/src/definition-json/main.go:19:8",
			"text": "\tvar x lib.T           // @definition lexical-pkgname \"lib\""
		},
		{
			"pos": "testdata/src/definition-json/main.go:25:8",
			"text": "\tvar _ lib.Type     // @definition qualified-type \"Type\""
		},
		{
			"pos": "testdata/src/definition-json/main.go:26:8",
			"text": "\tvar _ lib.Func     // @definition qualified-func \"Func\""
		},
		{
			"pos": "testdata/src/definition-json/main.go:27:8",
			"text": "\tvar _ lib.Var      // @definition qualified-var \"Var\""
		},
		{
			"pos": "testdata/src/definition-json/main.go:28:8",
			"text": "\tvar _ lib.Const    // @definition qualified-const \"Const\""
		},
		{
			"pos": "testdata/src/definition-json/main.go:29:8",
			"text": "\tvar _ lib2.Type    // @definition qualified-type-renaming \"Type\""
		},
		{
			"pos": "testdata/src/definition-json/main.go:30:8",
			"text": "\tvar _ lib.Nonesuch // @definition qualified-nomember \"Nonesuch\""
		},
		{
			"pos": "testdata/src/definition-json/main.go:63:2",
			"text": "\tlib.Type // @definition embedded-other-pkg \"Type\""
		}
	]
}
{
	"package": "describe",
	"refs": [
		{
			"pos": "testdata/src/describe/main.go:92:8",
			"text": "\tvar _ lib.Outer // @describe lib-outer \"Outer\""
		}
	]
}
{
	"package": "imports",
	"refs": [
		{
			"pos": "testdata/src/imports/main.go:18:12",
			"text": "\tconst c = lib.Const // @describe ref-const \"Const\""
		},
		{
			"pos": "testdata/src/imports/main.go:19:2",
			"text": "\tlib.Func()          // @describe ref-func \"Func\""
		},
		{
			"pos": "testdata/src/imports/main.go:20:2",
			"text": "\tlib.Var++           // @describe ref-var \"Var\""
		},
		{
			"pos": "testdata/src/imports/main.go:21:8",
			"text": "\tvar t lib.Type      // @describe ref-type \"Type\""
		},
		{
			"pos": "testdata/src/imports/main.go:26:8",
			"text": "\tvar _ lib.Type // @describe ref-pkg \"lib\""
		}
	]
}
{
	"package": "referrers",
	"refs": [
		{
			"pos": "testdata/src/referrers/int_test.go:7:7",
			"text": "\t_ = (lib.Type).Method // ref from internal test package"
		}
	]
}
{
	"package": "referrers",
	"refs": [
		{
			"pos": "testdata/src/referrers/main.go:16:8",
			"text": "\tvar v lib.Type = lib.Const // @referrers ref-package \"lib\""
		},
		{
			"pos": "testdata/src/referrers/main.go:16:19",
			"text": "\tvar v lib.Type = lib.Const // @referrers ref-package \"lib\""
		}
	]
}
{
	"package": "referrers-json",
	"refs": [
		{
			"pos": "testdata/src/referrers-json/main.go:14:8",
			"text": "\tvar v lib.Type = lib.Const // @referrers ref-package \"lib\""
		},
		{
			"pos": "testdata/src/referrers-json/main.go:14:19",
			"text": "\tvar v lib.Type = lib.Const // @referrers ref-package \"lib\""
		}
	]
}
{
	"package": "referrers_test",
	"refs": [
		{
			"pos": "testdata/src/referrers/ext_test.go:10:7",
			"text": "\t_ = (lib.Type).Method // ref from external test package"
		}
	]
}
{
	"package": "what-json",
	"refs": [
		{
			"pos": "testdata/src/what-json/main.go:13:7",
			"text": "var _ lib.Var // @what pkg \"lib\""
		},
		{
			"pos": "testdata/src/what-json/main.go:14:8",
			"text": "type _ lib.T"
		}
	]
}
-------- @referrers ref-method --------
{
	"objpos": "testdata/src/lib/lib.go:5:13",
	"desc": "func (lib.Type).Method(x *int) *int"
}
{
	"package": "imports",
	"refs": [
		{
			"pos": "testdata/src/imports/main.go:22:9",
			"text": "\tp := t.Method(\u0026a)   // @describe ref-method \"Method\""
		}
	]
}
{
	"package": "referrers",
	"refs": [
		{
			"pos": "testdata/src/referrers/int_test.go:7:17",
			"text": "\t_ = (lib.Type).Method // ref from internal test package"
		}
	]
}
{
	"package": "referrers",
	"refs": [
		{
			"pos": "testdata/src/referrers/main.go:17:8",
			"text": "\t_ = v.Method               // @referrers ref-method \"Method\""
		},
		{
			"pos": "testdata/src/referrers/main.go:18:8",
			"text": "\t_ = v.Method"
		}
	]
}
{
	"package": "referrers-json",
	"refs": [
		{
			"pos": "testdata/src/referrers-json/main.go:15:8",
			"text": "\t_ = v.Method               // @referrers ref-method \"Method\""
		},
		{
			"pos": "testdata/src/referrers-json/main.go:16:8",
			"text": "\t_ = v.Method"
		}
	]
}
{
	"package": "referrers_test",
	"refs": [
		{
			"pos": "testdata/src/referrers/ext_test.go:10:17",
			"text": "\t_ = (lib.Type).Method // ref from external test package"
		}
	]
}
-------- @referrers ref-local --------
{
	"objpos": "testdata/src/referrers-json/main.go:14:6",
	"desc": "var v lib.Type"
}
{
	"package": "referrers-json",
	"refs": [
		{
			"pos": "testdata/src/referrers-json/main.go:15:6",
			"text": "\t_ = v.Method               // @referrers ref-method \"Method\""
		},
		{
			"pos": "testdata/src/referrers-json/main.go:16:6",
			"text": "\t_ = v.Method"
		},
		{
			"pos": "testdata/src/referrers-json/main.go:17:2",
			"text": "\tv++ //@referrers ref-local \"v\""
		},
		{
			"pos": "testdata/src/referrers-json/main.go:18:2",
			"text": "\tv++"
		}
	]
}
-------- @referrers ref-field --------
{
	"objpos": "testdata/src/referrers-json/main.go:10:2",
	"desc": "field f int"
}
{
	"package": "referrers-json",
	"refs": [
		{
			"pos": "testdata/src/referrers-json/main.go:20:10",
			"text": "\t_ = s{}.f // @referrers ref-field \"f\""
		},
		{
			"pos": "testdata/src/referrers-json/main.go:23:5",
			"text": "\ts2.f = 1"
		}
	]
}
